client/db/src/utils.rs

```diff
@@ -449,10 +449,46 @@ impl DatabaseType {
 	}
 }
 
+
+pub(crate) struct JoinInput<I1, I2>(I1, I2);
+
+pub(crate) fn join_input<I1: codec::Input, I2: codec::Input>(i1: I1, i2: I2) -> JoinInput<I1, I2> {
+	JoinInput(i1, i2)
+}
+
+impl<I1: codec::Input, I2: codec::Input> codec::Input for JoinInput<I1, I2> {
+	fn remaining_len(&mut self) -> Result<Option<usize>, codec::Error> {
+		Ok(if let (Some(l1), Some(l2)) = (self.0.remaining_len()?, self.1.remaining_len()?) {
+			Some(l1 + l2)
+		} else {
+			None
+		})
+	}
+
+	fn read(&mut self, into: &mut [u8]) -> Result<(), codec::Error> {
+		let mut read = 0;
+		match self.0.remaining_len()? {
+			Some(l) if l == 0 => {},
+			Some(l) => {
+				read = std::cmp::min(l, into.len());
+				self.0.read(&mut into[..read])?;
+			}
+			None => {
+				return self.0.read(into)
```

Comment:
This means we will never read `self.1`. For our use case here it is okay, because we always should get `Some()`, but that should really not be used by someone else.

Maybe we should use `unreachable` here or make `JoinInput` directly use `Vec<u8>` and `Option<Vec<u8>>`

Reply:
Changed it to operate on slices instead